logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -0,0 +1,388 @@
+            child -> {
+                final Collection<Dataset> newparents = flatten(parents, child);
+                if (isFilter(child)) {
+                    return filterDataset(child.getId(), newparents);
+                } else if (isOutput(child)) {
+                    return outputDataset(child.getId(), newparents);
+                } else {
+                    // We know that it's an if vertex since the the input children are either 
+                    // output, filter or if in type.
+                    final IfVertex ifvert = (IfVertex) child;
+                    final EventCondition iff = buildCondition(ifvert);
+                    final String index = ifvert.getId();
+                    // It is important that we double check that we are actually dealing with the
+                    // positive/left branch of the if condition
+                    if (ifvert.getOutgoingBooleanEdgesByType(true).stream()
+                        .anyMatch(edge -> Objects.equals(edge.getTo(), start))) {
+                        return splitLeft(newparents, iff, index);
+                    } else {
+                        return splitRight(newparents, iff, index);
+                    }
+                }
+            }).collect(Collectors.toList());
+    }
+
+    /**
+     * Build a {@link Dataset} representing the {@link JrubyEventExtLibrary.RubyEvent}s after
+     * the application of the given filter.
+     * @param vertex Vertex Id of the filter to create this {@link Dataset} for
+     * @param parents All the parent nodes that go into this filter
+     * @return Filter {@link Dataset}
+     */
+    private Dataset filterDataset(final String vertex, final Collection<Dataset> parents) {
+        final Dataset filter;
+        if (plugins.containsKey(vertex)) {
+            filter = plugins.get(vertex);
+        } else {
+            final RubyIntegration.Filter ruby = filters.get(vertex);
+            if (ruby.hasFlush()) {
+                if (ruby.periodicFlush()) {
+                    filter = new Dataset.FilteredFlushableDataset(parents, ruby);
+                } else {
+                    filter = new Dataset.FilteredShutdownFlushableDataset(parents, ruby);
+                }
+            } else {
+                filter = new Dataset.FilteredDataset(parents, ruby);
+            }
+            plugins.put(vertex, filter);
+        }
+        return filter;
+    }
+
+    /**
+     * Build a {@link Dataset} representing the {@link JrubyEventExtLibrary.RubyEvent}s after
+     * the application of the given output.
+     * @param vertex Vertex Id of the filter to create this {@link Dataset} for
+     * filter node in the topology once
+     * @param parents All the parent nodes that go into this output
+     * @return Output {@link Dataset}
+     */
+    private Dataset outputDataset(final String vertex, final Collection<Dataset> parents) {
+        final Dataset output;
+        if (plugins.containsKey(vertex)) {
```

Comment:
This block of code would be a great place to use `Map.computeIfAbsent` for clarity.